src/searching/saddleback_search.rs

```diff
@@ -1,85 +1,196 @@
-// Saddleback search is a technique used to find an element in a sorted 2D matrix in O(m + n) time,
-// where m is the number of rows, and n is the number of columns. It works by starting from the
-// top-right corner of the matrix and moving left or down based on the comparison of the current
-// element with the target element.
 use std::cmp::Ordering;
 
-pub fn saddleback_search(matrix: &[Vec<i32>], element: i32) -> (usize, usize) {
-    // Initialize left and right indices
+/// Custom error type to represent errors related to matrix validation.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MatrixError {
+    NonRectangularMatrix,
+}
+
+/// Performs Saddleback search on a sorted 2D matrix.
+///
+/// The Saddleback search algorithm finds the position of a target element in a matrix where
+/// each row and each column is sorted in ascending order. The search starts from the top-right
+/// corner of the matrix and moves left or down based on comparisons with the target element.
+///
+/// # Arguments
+///
+/// * `matrix` - A 2D vector representing the sorted matrix.
```

Comment:
2D vector is a bit misleading here. Maybe an expression like _vector of vectors_ would be better? Or _2D array_?

Reply:
Referring to `matrix` as a _2D vector_ can be a bit confusing, especially since Rust doesn't have native 2D arrays like some other languages. In Rust, the type we are working with is a vector of vectors (`Vec<Vec<isize>>`), which represents a matrix-like structure but isn't truly a 2D array.